sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -97,6 +98,24 @@ case class BroadcastNestedLoopJoinExec(
     }
   }
 
+
+  /**
+   * See. [SPARK-32290]
+   * Not in Subquery will almost certainly be planned as a Broadcast Nested Loop join
+   * which is very time consuming because it's a M*N calculation. But if it's a single column
+   * NotInSubquery, and buildSide data is small enough, M*N calculation could be optimized into
+   * M*log(N) using HashMap to lookup build side rows.
+   * Meet following condition the optimize will works:
+   * NOT_IN_SUBQUERY_SINGLE_COLUMN_OPTIMIZE_ENABLED is true
+   * BuildSideRowCount <= NOT_IN_SUBQUERY_SINGLE_COLUMN_OPTIMIZE_ROW_COUNT_THRESHOLD
+   */
+
+  // NotInSubquery Key in StreamedSide
+  private var notInSubquerySingleColumnOptimizeStreamedKey: Attribute = _
```

Comment:
I am wondering if it is possible to rework the code so that we are not passing "return values using field variables" ? Its kind of hacky to do this in general :-) 

Can notInSubquerySingleColumnOptimizeStreamedKey, notInSubquerySingleColumnOptimizeStreamedKeyIndex for example be returned as perhaps by notInSubquerySingleColumnOptimizeEnabled, perhaps by having it return an Option[(Attribute, Int)] which is None if the optimization cannot be enabled.